File: EOCV-Sim/src/main/java/com/github/serivesmejia/eocvsim/gui/dialog/Configuration.java
/*
 * Copyright (c) 2021 Sebastian Erives
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package com.github.serivesmejia.eocvsim.gui.dialog;

import com.github.serivesmejia.eocvsim.EOCVSim;
import com.github.serivesmejia.eocvsim.config.Config;
import com.github.serivesmejia.eocvsim.gui.component.input.EnumComboBox;
import com.github.serivesmejia.eocvsim.gui.component.input.SizeFields;
import com.github.serivesmejia.eocvsim.gui.theme.Theme;
import com.github.serivesmejia.eocvsim.gui.util.WebcamDriver;
import com.github.serivesmejia.eocvsim.pipeline.PipelineFps;
import com.github.serivesmejia.eocvsim.pipeline.PipelineTimeout;

import javax.swing.*;
import java.awt.*;

public class Configuration {

    private final EOCVSim eocvSim;
    public JPanel contents = new JPanel(new GridBagLayout());
    public JComboBox<String> themeComboBox = new JComboBox<>();

    public JButton acceptButton = new JButton("Accept");

    public JCheckBox pauseOnImageCheckBox = new JCheckBox();

    public EnumComboBox<PipelineTimeout> pipelineTimeoutComboBox = null;
    public EnumComboBox<PipelineFps> pipelineFpsComboBox = null;

    public EnumComboBox<WebcamDriver> preferredWebcamDriver = null;

    public SizeFields videoRecordingSize = null;
    public EnumComboBox<PipelineFps> videoRecordingFpsComboBox = null;

    JDialog configuration;

    public Configuration(JFrame parent, EOCVSim eocvSim) {
        configuration = new JDialog(parent);
        this.eocvSim = eocvSim;

        eocvSim.visualizer.childDialogs.add(configuration);

        initConfiguration();
    }

    private void initConfiguration() {

        Config config = this.eocvSim.configManager.getConfig();
        configuration.setModal(true);
        configuration.setTitle("Settings");
        //configuration.setSize(380, 300);

        JTabbedPane tabbedPane = new JTabbedPane(JTabbedPane.LEFT);

        /*
        UI TAB
         */

        JPanel uiPanel = new JPanel(new GridLayout(1, 1, 1, 8));

        /* THEME SETTING */
        JPanel themePanel = new JPanel(new FlowLayout());
        JLabel themeLabel = new JLabel("Theme: ");

        themeLabel.setHorizontalAlignment(0);

        for (Theme theme : Theme.values()) {
            this.themeComboBox.addItem(theme.toString().replace("_", " "));
        }

        themeComboBox.setSelectedIndex(eocvSim.getConfig().simTheme.ordinal());
        themePanel.add(themeLabel);
        themePanel.add(this.themeComboBox);
        uiPanel.add(themePanel);

        tabbedPane.addTab("Inteface", uiPanel);

        /*
        INPUT SOURCES TAB
         */
        JPanel inputSourcesPanel = new JPanel(new GridLayout(2, 1, 1, 8));

        /* PAUSE WITH IMAGE SOURCES OPTION */
        JPanel pauseOnImagePanel = new JPanel(new FlowLayout());
        JLabel pauseOnImageLabel = new JLabel("Pause with Image Sources");

        pauseOnImageCheckBox.setSelected(config.pauseOnImages);

        pauseOnImagePanel.add(pauseOnImageCheckBox);
        pauseOnImagePanel.add(pauseOnImageLabel);

        inputSourcesPanel.add(pauseOnImagePanel);

        /* PREFERRED WEBCAM DRIVER OPTION */
        preferredWebcamDriver = new EnumComboBox<>(
                "Preferred Webcam Driver: ", WebcamDriver.class,
                WebcamDriver.values()
        );
        preferredWebcamDriver.setSelectedEnum(config.preferredWebcamDriver);
        inputSourcesPanel.add(preferredWebcamDriver);

        tabbedPane.addTab("Input Sources", inputSourcesPanel);

        /*
        PROCESSING TAB
         */
        JPanel processingPanel = new JPanel(new GridLayout(4, 1, 1, 8));

        /* PIPELINE TIMEOUT IN processFrame AND init*/
        pipelineTimeoutComboBox = new EnumComboBox<>(
                "Pipeline Process Timeout: ", PipelineTimeout.class,
                PipelineTimeout.values(), PipelineTimeout::getCoolName, PipelineTimeout::fromCoolName
        );
        pipelineTimeoutComboBox.setSelectedEnum(config.pipelineTimeout);
        processingPanel.add(pipelineTimeoutComboBox);

        /* PIPELINE FPS*/
        pipelineFpsComboBox = new EnumComboBox<>(
                "Pipeline Max FPS: ", PipelineFps.class,
                PipelineFps.values(), PipelineFps::getCoolName, PipelineFps::fromCoolName
        );
        pipelineFpsComboBox.setSelectedEnum(config.pipelineMaxFps);
        processingPanel.add(this.pipelineFpsComboBox);

        /* VIDEO REC SIZE */
        videoRecordingSize = new SizeFields(
                config.videoRecordingSize, false,
                "Video Recording Size: "
        );
        videoRecordingSize.onChange.doPersistent(() ->
                acceptButton.setEnabled(this.videoRecordingSize.getValid())
        );
        processingPanel.add(this.videoRecordingSize);

        /* VIDEO REC FPS */
        videoRecordingFpsComboBox = new EnumComboBox<>(
                "Video Recording FPS: ", PipelineFps.class,
                PipelineFps.values(), PipelineFps::getCoolName, PipelineFps::fromCoolName
        );
        videoRecordingFpsComboBox.setSelectedEnum(config.videoRecordingFps);
        processingPanel.add(videoRecordingFpsComboBox);

        tabbedPane.addTab("Processing", processingPanel);

        GridBagConstraints gbc = new GridBagConstraints();

        /* ADD TABBED PANE TO DIALOG */
        gbc.gridx = 1;
        gbc.gridy = 1;
        contents.add(tabbedPane, gbc);

        /*
        ACCEPT BUTTON AND FINISH SETTING UP DIALOG
         */
        JPanel acceptPanel = new JPanel(new FlowLayout());
        acceptPanel.add(acceptButton);

        acceptButton.addActionListener(e -> {
            this.eocvSim.onMainUpdate.doOnce(this::applyChanges);
            close();
        });

        gbc.gridy = 2;
        contents.add(acceptPanel, gbc);
        contents.setBorder(BorderFactory.createEmptyBorder(10, 0, 10, 0));

        configuration.add(this.contents);
        configuration.pack();

        configuration.setResizable(false);
        configuration.setLocationRelativeTo(null);
        configuration.setVisible(true);
    }

    private void applyChanges() {
        Config config = eocvSim.configManager.getConfig();

        Theme userSelectedTheme = Theme.valueOf(themeComboBox.getSelectedItem().toString().replace(" ", "_"));
        Theme beforeTheme = config.simTheme;

        //save user modifications to config
        config.simTheme = userSelectedTheme;
        config.pauseOnImages = pauseOnImageCheckBox.isSelected();
        config.preferredWebcamDriver = preferredWebcamDriver.getSelectedEnum();
        config.pipelineTimeout = pipelineTimeoutComboBox.getSelectedEnum();
        config.pipelineMaxFps = pipelineFpsComboBox.getSelectedEnum();
        config.videoRecordingSize = videoRecordingSize.getCurrentSize();
        config.videoRecordingFps = videoRecordingFpsComboBox.getSelectedEnum();

        eocvSim.configManager.saveToFile(); //update config file

        if (userSelectedTheme != beforeTheme)
            eocvSim.restart();
    }

    public void close() {
        configuration.setVisible(false);
        configuration.dispose();
    }

}

File: EOCV-Sim/src/main/java/com/github/serivesmejia/eocvsim/gui/dialog/Configuration.kt
/*
 * Copyright (c) 2021 Sebastian Erives
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package com.github.serivesmejia.eocvsim.gui.dialog

import com.github.serivesmejia.eocvsim.LifecycleSignal
import com.github.serivesmejia.eocvsim.config.ConfigManager
import com.github.serivesmejia.eocvsim.gui.DialogFactory
import com.github.serivesmejia.eocvsim.gui.Visualizer
import com.github.serivesmejia.eocvsim.gui.component.input.EnumComboBox
import com.github.serivesmejia.eocvsim.gui.component.input.SizeFields
import com.github.serivesmejia.eocvsim.gui.theme.Theme
import com.github.serivesmejia.eocvsim.pipeline.PipelineFps
import com.github.serivesmejia.eocvsim.pipeline.PipelineTimeout
import com.github.serivesmejia.eocvsim.util.event.EventHandler
import kotlinx.coroutines.channels.Channel
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import org.koin.core.qualifier.named
import java.awt.FlowLayout
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.awt.GridLayout
import javax.swing.*

class Configuration : KoinComponent {

    private val visualizer: Visualizer by inject()
    private val lifecycle: Channel<LifecycleSignal> by inject(named("lifecycle"))
    private val dialogFactory: DialogFactory by inject()
    private val configManager: ConfigManager by inject()
    private val onMainLoop: EventHandler by inject(named("onMainLoop"))

    private val config get() = configManager.config

    private val dialog = JDialog(visualizer.frame)

    private val themeComboBox: JComboBox<String>
    private val superAccessCheckBox: JCheckBox
    private val prefersPaperVisionCheckbox: JCheckBox
    private val pauseOnImageCheckBox: JCheckBox
    private val webcamOpenTimeoutSpinner: JSpinner
    private val webcamNewFrameTimeoutSpinner: JSpinner
    private val pipelineTimeoutComboBox: EnumComboBox<PipelineTimeout>
    private val pipelineFpsComboBox: EnumComboBox<PipelineFps>
    private val videoRecordingSize: SizeFields
    private val videoRecordingFpsComboBox: EnumComboBox<PipelineFps>
    private val acceptButton: JButton

    init {
        // --- Interface Tab ---
        themeComboBox = JComboBox<String>().apply {
            Theme.entries.forEach { addItem(it.name.replace("_", " ")) }
            selectedIndex = config.simTheme.ordinal
        }
        superAccessCheckBox = JCheckBox("Auto Accept SuperAccess on Trusted Plugins").apply {
            isSelected = config.autoAcceptSuperAccessOnTrusted
        }
        prefersPaperVisionCheckbox = JCheckBox("Focus on PaperVision Upon Startup").apply {
            isSelected = config.flags["prefersPaperVision"] ?: false
        }
        val uiPanel = JPanel(GridLayout(3, 1, 1, 8)).apply {
            add(JPanel(FlowLayout()).apply {
                add(JLabel("Theme: "))
                add(themeComboBox)
            })
            add(JPanel(FlowLayout()).apply { add(superAccessCheckBox) })
            add(JPanel(FlowLayout()).apply { add(prefersPaperVisionCheckbox) })
        }

        // --- Input Sources Tab ---
        pauseOnImageCheckBox = JCheckBox("Pause with Image Sources").apply {
            isSelected = config.pauseOnImages
        }
        webcamOpenTimeoutSpinner = JSpinner(SpinnerNumberModel(config.webcamOpenTimeoutSec, 0.1, 60.0, 0.1))
        webcamNewFrameTimeoutSpinner = JSpinner(SpinnerNumberModel(config.webcamNewFrameTimeoutSec, 0.1, 60.0, 0.1))
        val inputSourcesPanel = JPanel(GridLayout(3, 1, 1, 8)).apply {
            add(JPanel(FlowLayout()).apply { add(pauseOnImageCheckBox) })
            add(JPanel(FlowLayout()).apply {
                add(JLabel("Timeout to start camera stream (seconds): "))
                add(webcamOpenTimeoutSpinner)
            })
            add(JPanel(FlowLayout()).apply {
                add(JLabel("Timeout for new camera frame (seconds): "))
                add(webcamNewFrameTimeoutSpinner)
            })
        }

        // --- Processing Tab ---
        pipelineTimeoutComboBox = EnumComboBox(
            "Pipeline Process Timeout: ",
            PipelineTimeout::class.java,
            PipelineTimeout.entries.toTypedArray(),
            PipelineTimeout::coolName
        ) { PipelineTimeout.fromCoolName(it) ?: PipelineTimeout.MEDIUM }.apply { selectedEnum = config.pipelineTimeout }

        pipelineFpsComboBox = EnumComboBox(
            "Pipeline Max FPS: ",
            PipelineFps::class.java,
            PipelineFps.entries.toTypedArray(),
            PipelineFps::coolName
        ) { PipelineFps.fromCoolName(it) ?: PipelineFps.MEDIUM }.apply { selectedEnum = config.pipelineMaxFps }

        acceptButton = JButton("Accept") // Initialized here so the listener below can use it

        videoRecordingSize = SizeFields(
            config.videoRecordingSize,
            false,
            "Video Recording Size: "
        ).apply { onChange.attach { acceptButton.isEnabled = valid } }

        videoRecordingFpsComboBox = EnumComboBox(
            "Video Recording FPS: ",
            PipelineFps::class.java,
            PipelineFps.entries.toTypedArray(),
            PipelineFps::coolName
        ) { PipelineFps.fromCoolName(it) ?: PipelineFps.MEDIUM }.apply { selectedEnum = config.videoRecordingFps }

        val processingPanel = JPanel(GridLayout(4, 1, 1, 8)).apply {
            add(pipelineTimeoutComboBox)
            add(pipelineFpsComboBox)
            add(videoRecordingSize)
            add(videoRecordingFpsComboBox)
        }

        // --- Dialog Assembly ---
        val tabbedPane = JTabbedPane(JTabbedPane.LEFT).apply {
            addTab("Interface", uiPanel)
            addTab("Input Sources", inputSourcesPanel)
            addTab("Processing", processingPanel)
        }

        acceptButton.addActionListener {
            onMainLoop.once {
                applyChanges()
            }

            close()
        }

        val contents = JPanel(GridBagLayout()).apply {
            border = BorderFactory.createEmptyBorder(10, 0, 10, 0)
            val gbc = GridBagConstraints()

            gbc.gridx = 1; gbc.gridy = 1
            add(tabbedPane, gbc)

            gbc.gridy = 2
            add(JPanel(FlowLayout()).apply { add(acceptButton) }, gbc)
        }

        dialog.apply {
            add(contents)
            title = "Settings"
            isModal = true
            isResizable = false
            pack()
            setLocationRelativeTo(null)
            isVisible = true
        }
    }

    private fun applyChanges() {
        val selectedThemeName = themeComboBox.selectedItem!!.toString().replace(" ", "_")
        val userSelectedTheme = Theme.valueOf(selectedThemeName)
        val previousTheme = config.simTheme

        config.simTheme = userSelectedTheme
        config.pauseOnImages = pauseOnImageCheckBox.isSelected
        config.webcamOpenTimeoutSec = (webcamOpenTimeoutSpinner.value as Number).toDouble()
        config.webcamNewFrameTimeoutSec = (webcamNewFrameTimeoutSpinner.value as Number).toDouble()
        config.pipelineTimeout = pipelineTimeoutComboBox.selectedEnum
        config.pipelineMaxFps = pipelineFpsComboBox.selectedEnum
        config.videoRecordingSize = videoRecordingSize.currentSize
        config.videoRecordingFps = videoRecordingFpsComboBox.selectedEnum
        config.autoAcceptSuperAccessOnTrusted = superAccessCheckBox.isSelected
        config.flags["prefersPaperVision"] = prefersPaperVisionCheckbox.isSelected

        configManager.saveToFile()

        if (userSelectedTheme != previousTheme) {
            dialogFactory.createYesOrNo(dialog, "Applying a new interface theme requires restarting.", "Do you wish to restart now?") {
                lifecycle.trySend(LifecycleSignal.Restart)
            }
        }
    }

    fun close() {
        dialog.isVisible = false
        dialog.dispose()
    }
}